### DISCUSSION TITLE
[Feature Request] Prevent issues from being closed by merging linked PRs

### DISCUSSION THREAD

**Discussion by akshaymankar** [community/community#23476] [Pull Requests]
<p>I really like how the linked PRs in an issue are listed nicely on the project board as well as in the right side of the issue. It really makes discovering PRs from issues easy.</p>
<p>In my current project, we don’t close issues until they reach production. Also, often we have multiple PRs which fix an issues (in multiple repositories). In such cases I do not want the issue to be autmatically closed, currently I just re-open them. This is not ideal because I could forget to re-open the issue and also this workflow causes a lot of notification noise in my team.</p>
<p>It would be nice if this could somehow be configured, or if there was another keyword (instaed of fixes, closes, etc.) to signify that a PR only partially fixes an issue and the issue and the issue must not be closed auotmatically even after the PR is merged.</p>

**Comment by AndreaGriffiths11** [comment, marked as answer]
<p>Hi <a class="mention" href="/akshaymankar">@akshaymankar</a>,</p>
<p>Thanks for this feedback! We’re always working to improve GitHub, and we consider every suggestion we receive. I’ve located an existing “feature request” for disabling the auto-close issues feature and added your voice to it.</p>
<p>Though I can’t guarantee anything or share a timeline for this, I can tell you that it’s been shared with the appropriate teams for consideration.</p>

**Reply by dariusonsched** [reply]
@AndreaGriffiths11 it's time to get this feature done.  Look at all of the support and upvotes in this ticket.  2+ years is embarrassingly slow.

**Reply by kingthorin** [reply]
Ya look at all the support, there's like 200 people that +1ed out of your user base of millions........ :grinning: (how many zeros are in that decimal?) Prioritize this thing that people couldn't even be bothered to write more than two characters about.

Don't get me wrong, I understand there are people that want this. I just find the tail trying to wag the dog really REALLY funny :grinning:

**Reply by dariusonsched** [reply]
@kingthorin While I appreciate you being entitled to your opinion, nothing I said in my post solicited a snarky response.  There are actual enterprises out there trying to use GitHub with an Enterprise subscription and this functionality - which exists within Azure DevOps, does not exist here (_yet_).  It's time it was.  Have you worked with a project board and workflow within GitHub?  Our workaround uses a bot to re-open the Issue/ticket upon moving it across the board following a PR/merge review because it has not yet gone through QA and UAT.  An issue is not considered Closed until it ships.

**Reply by kingthorin** [reply]
> While I appreciate you being entitled to your opinion, nothing I said in my post solicited a snarky response. 

You're right and that's fair. Sorry the "smart a$$" was being strong in me this morning :wink:

> There are actual enterprises out there trying to use GitHub with an Enterprise subscription and this functionality - which exists within Azure DevOps, does not exist here (yet).

While I appreciate that GitHub is now owned by MS so this might seem like a Apples to Apples comparison it really isn't, and that should be obvious.

> It's time it was.

That's easy to decide with no insight or forgiveness. None of us are in a position to say how easy that is or isn't or really how important it is to the roadmap/devs/mgmt/exec.

> Have you worked with a project board and workflow within GitHub? 

Yup, but all projects do this differently.

> Our workaround uses a bot to re-open the Issue/ticket upon moving it across the board following a PR/merge review because it has not yet gone through QA and UAT. An issue is not considered Closed until it ships.

I can see that kinda sucks, of course it would be nice if it was a native solution. But you're functional, though agreeably sub-optimally.

---

I'm sorry I found the whole thing humorous earlier. But I think articulating your pain (as you've now done) is 1000x more meaningful than just another "+1" or "this sucks why haven't you done it yet".

**Reply by dariusonsched** [reply]
@kingthorin No hard feelings and I appreciate your follow-up.  Just wanted you to be aware, I had already _previously_ shared back in January, with proper articulation, why this issue should be addressed and have been following the topic.

Link to my initial feedback on this issue:
https://github.com/orgs/community/discussions/23476#discussioncomment-3240520

Also, there is a lot of synergy in terms of cross-feature application from the Azure DevOps solution into GitHub for the enterprise/corporate user base.  Colleagues of mine at MSFT have informed me of same. 

Something is already auto-closing the Issue when a linked PR is completed, therefore it's already an automated process / RPA.  Therefore, a simple setting/feature flag to disable this from occurring seems straight forward.  But, I'm just a software architect with over 25 years of real world experience.  This GH profile is not my primary.  It's just one I use while helping a client.

**Reply by misabitencourt** [reply]
@AndreaGriffiths11  i would really appreciate if this feature would be released.

**Reply by bitdivine** [reply]
@AndreaGriffiths11 - Do you have a link to the feature request you mentioned, so that we can see how it is progressing?

**Reply by AndreaGriffiths11** [reply]
Thank you for reaching out and for the ping, @bitdivine. I apologize for the oversight in missing the notification. I switched teams, and my notifications have become a bit disorganized (TBH Chaotic is a better word). Our feature request log is not publicly accessible. However, I have located the issue and requested an update. I will promptly share the information here once I receive it.

**Reply by marc-hb** [reply]
> Yup, but all projects do this differently.

... exactly why forcing all projects to auto-close is so bad. It must be a choice.

> None of us are in a position to say how easy that is 

While it's obviously hard to tell the actual complexity considering it's all closed-source, only a _boolean_ is being requested in terms of user-visible effects. It would be very surprising and disappointing if that required very complex changes and a huge development effort.

>  there's like 200 people that +1ed out of your user base of millions.

(it's close to 350 now)

Upvoting is not "natural", I've seen issues with more "me too" comments that upvotes. It also requires caring enough to look for an existing issue and being skilled and persistent enough to find it; that's a pretty high bar.

The more interesting metric is to compare with how many votes _other_ issues got.

**Reply by bitdivine** [reply]
> However, I have located the issue and requested an update.

@AndreaGriffiths11 Thank you.  Was any update provided?

**Reply by caguilar3** [reply]
Me and my team really needs this <3 <3

**Reply by ChristianTellefsen** [reply]
I need the option to not have issues auto-close when a single linked PR is closed.

**Reply by JaKoutny** [reply]
@AndreaGriffiths11 Sorry to ping you, but you promised us prompt information more than six months ago. It would be nice to at least know that someone is dealing with this (or not).

**Reply by janwerkhoven** [reply]
4 years later ... STILL AN ISSUE 🔥

**Reply by larissapissurno** [reply]
any update? 😢

**Reply by lidiaparrilla** [reply]
This issue is closed!! Github will not pay any attention to this. Move over to this follow-up: https://github.com/orgs/community/discussions/66741

**Reply by marc-hb** [reply]
Just FYI there's another duplicate in 2022 #17308. Filed later but with a lower number; go figure.

**Reply by Integer93** [reply]
@AndreaGriffiths11 is this still being worked on?

**Reply by anne-gropler** [reply]
So annoying that if we write in a PR description "This does not fix #123456789 because foo bar", that issue will be linked as "fixes #123456789"....

**Reply by marc-hb** [reply]
https://github.blog/changelog/2025-04-23-users-can-now-choose-whether-merging-linked-pull-requests-automatically-closes-the-issue/

**Comment by GonzaloCalandria** [comment]
<p>Was this solved? I’ve the same problem and I would like to know if this feature request was applied or where can I see this feature request.</p>

**Reply by marc-hb** [reply]
https://github.blog/changelog/2025-04-23-users-can-now-choose-whether-merging-linked-pull-requests-automatically-closes-the-issue/

**Comment by rkrishna-gembrill** [comment]
<p>I really expect this feature as it is one primary use case where developer fix has to be verified by QA</p>

**Reply by Lyokolux** [reply]
Indeed

**Reply by marc-hb** [reply]
QA what? /s

**Comment by ahmed2m** [comment]
<p>This should be made with another keyword like for instance ‘Linked to <span class="hashtag">#123</span>’</p>

**Comment by kingthorin** [comment]
<p>You can link issues without using an auto-close keyword. Just use <code>org/repo#issue_number</code> Ex: <a href="https://github.com/OWASP/OWASP-VWAD/issues/1" rel="noopener nofollow ugc">OWASP/OWASP-VWAD#1</a></p>
<p>I guess technically you won’t see them in that little “Linked issues” area in the PRs but you’ll still see the reference in the PR message and in the Issue history/log.</p>

**Comment by 3dfoster** [comment]
<p>Hi <a class="mention" href="/AndreaGriffiths11">@AndreaGriffiths11</a>,</p>
<p>Can you provide a link to the existing feature request?</p>
<p>Thanks.</p>

**Reply by marc-hb** [reply]
Or at least a reference number.

**Comment by bitdivine** [comment]
<p>Not using a github keyword such as <code>fix</code> is not a viable option if the commit text format is already specified and well established.  An option to disable auto-closing would be preferable.</p>

**Comment by hipersayanX** [comment]
<p>Agree, issues should never be closed before a release, since the issue still exist in the latest release, and it can take a long time before a new release with the fix applied is made.</p>

**Comment by amnonkhen** [comment]
<p>Has there been any progress on this issue?<br>
Can someone please update the feature request link or status?<br>
Thanks.</p>

**Comment by JSoet** [comment]
<p><a class="mention" href="/AndreaGriffiths11">@AndreaGriffiths11</a><br>
You mentioned you located an existing feature request, do you have a link to that feature request so we can track the status and add our voices as well? Neither I or the other members who’ve replied here have found it.</p>

**Comment by alacret** [comment]
<p>any news here?</p>
<p>I really like how Kanban shows the linked PRs. But we can use this as it closes the issue</p>

**Comment by didac-adria** [comment]
<p>Hello!</p>
<p>We have kind of similar issue. In our case it’s ok for issues to be closed when there is ONLY one PR. However, when having multiple linked PR’s, we expect the issue to be closed when the last open PR is merged.</p>
<p>Thanks github team for your effort on continuous improvement!</p>

**Comment by demisx** [comment]
<p>We need an ability to auto-close most issues and then not to auto-close others. So, a separate keyword similar to <code>"Linked #&lt;issue_number&gt;"</code> or  <code>"Related #&lt;issue_number&gt;"</code> will be preferred.</p>

**Comment by dariusonsched** [comment]
<p>Hello <a class="mention" href="/andreagriffiths11">@AndreaGriffiths11</a>,</p>
<p>The community at large both here on this thread and throughout GitHub would substantially benefit from this feature.  Would you please share a link to the feature request you speak of for disabling auto-close issues so we can add our voices to it?</p>
<p>GitHub needs to take this issue seriously.  I have already had to move several projects to Azure DevOps as this behavior is not enterprise-worthy.  Closing an issue when a correlated Pull Request is merged into the default (typically integration/develop branch) is an incorrect workflow (for most projects).  An issue is not closed until it has been tested and slated for deployment (or even deployed) into production or cut into a release.</p>
<p>Thank you,<br>
Darius</p>

**Comment by ZachCMP** [comment]
<p>Hey <a class="mention" href="/andreagriffiths11">@AndreaGriffiths11</a>,</p>
<p>Can you please provide a link to the feature request and/or and update on this issue?  As with others on this thread, this behavior is problematic for my company’s workflow also and some ability to control it would be really helpful.</p>
<p>Thanks,</p>
<p>Zach</p>

**Comment by zwass** [comment]
<p><a class="mention" href="/andreagriffiths11">@AndreaGriffiths11</a> our team also has trouble with this. We don’t want our issues to mark as closed until we create a release with the merged pull requests.</p>

**Comment by spalfy** [comment]
<p>+1. Would be so simple.<br>
Also, why is this marked as solved?</p>

**Comment by mgsource7** [comment]
<p>+1 - also looking for this feature.</p>

**Comment by chrisduong** [comment]
<p>+1 - also desperately needs this!</p>

**Comment by roman-tiukh** [comment]
+1

**Comment by berkanteber** [comment]
+1

**Comment by joshlevy** [comment]
+1

**Comment by zarifmahfuz** [comment]
+1

**Comment by KristenPire** [comment]
+1

**Comment by duffbeeeer** [comment]
+1

**Comment by tianscenius** [comment]
+1

**Comment by kingthorin** [comment]
So many insightful +1s. You know you could simply vote up one of the posts right?

**Comment by alishaevn** [comment]
this question was duplicated (and is currently unanswered) in the "issues" category as well.
https://github.com/orgs/community/discussions/17308

**Comment by matthprost** [comment]
+1

**Comment by omidmaldar** [comment]
waiting for this feature to be released.

**Comment by bamorim** [comment]
+1

**Comment by doneitas** [comment]
+1

**Comment by tryforceful** [comment]
+1

**Comment by vicentllongo** [comment]
+1

**Comment by twencl** [comment]
+1

**Reply by Birdza87** [reply]
> +1
>

**Comment by gpressutto5** [comment]
+1

**Comment by jakemhill** [comment]
+1

**Comment by hwc0919** [comment]
+1

**Comment by jordanmiguel** [comment]
+1

**Comment by stephanhirsch** [comment]
+1

**Comment by smatting** [comment]
+1

**Comment by thewebprincess** [comment]
+1

**Comment by mgkeeley** [comment]
+1

**Comment by Strizzi12** [comment]
+1

**Comment by rkottmann** [comment]
+1

**Comment by bdewein** [comment]
+1

**Comment by domenic-donato** [comment]
Issues like this are why big companies switch off of GitHub at some point.

**Comment by hianhianhian** [comment]
+1

**Comment by lidiaparrilla** [comment]
+1

**Comment by jacksmith15** [comment]
I have implemented a workaround for this using GitHub Actions:

<details><summary><b>Click to view GH Action source</b></summary>

```yaml
name: Override default linked PR auto-closing
on:
  issues:
    types:
      - closed
jobs:
  reopen_autoclosed_issues:
    runs-on: ubuntu-latest
    steps:
      - name: Re-open autoclosed issues
        env:
          ORGANIZATION: my-org
          GITHUB_TOKEN: ${{secrets.GITHUB_TOKEN}}
          ISSUE_NUMBER: ${{github.event.issue.number}}
        run: |
          gh api graphql -f query='query($org: String!, $number: Int!) {
            organization(login: $org) {
              repository(name: "my-repository") {
                issue(number: $number) {
                  id
                  number
                  title
                  timelineItems(itemTypes: [CLOSED_EVENT], last: 1) {
                    __typename
                    ... on IssueTimelineItemsConnection {
                      nodes {
                        ... on ClosedEvent {
                          id
                          createdAt
                          closer {
                            __typename
                            ... on PullRequest {
                              url
                            }
                          }
                        }
                      }
                    }
                  }
                }
              }
            }
          }' -f org=$ORGANIZATION -F number=$ISSUE_NUMBER > issue-closed-event.json

          if $(jq '.data.organization.repository.issue.timelineItems.nodes[0].closer.__typename == "PullRequest"' issue-closed-event.json)
          then
            echo "Issue was closed by pull request: $(jq -r '.data.organization.repository.issue.timelineItems.nodes[0].closer.url' issue-closed-event.json)"
            gh issue reopen ${{github.event.issue.html_url}}
            echo "Re-opened issue ${{github.event.issue.html_url}}"
          else
            echo "Issue was not closed by pull request, skipping"
          fi

```

</details>

This detects when an issue was closed because of a linked pull request, and re-opens the issue. The issue will still have "issue closed" followed by "issue re-opened" in its history, but its generally pretty quick to re-open issues.

It doesn't do anything to cards on project boards - but you can easily set up automation on a Project the move issues out of the "Done" column when they are re-opened.

**Reply by lidiaparrilla** [reply]
This is the world upside down. Having to implement a custom action to prevent an auto action instead of having and opt-out or even making this opt-in in the project settings, which would make more sense for me. Anyway, thanks @jacksmith15, good stuff.

**Comment by fxmarty** [comment]
![ezgif com-gif-maker](https://user-images.githubusercontent.com/9808326/207027474-2ffde812-c8d5-4dc5-979d-43da41461139.gif)

Github, why not stopping adding bloated features (as hijacking the user's ctrl + F) and working on actually useful stuff?

**Reply by mohsin** [reply]
Because...
![well-well-well](https://user-images.githubusercontent.com/1699753/211129363-0c72eddd-6163-4f65-aea1-9f51dfbd0505.jpg)

**Comment by uyarn** [comment]
+1

**Comment by afterburn** [comment]
+1

**Comment by mo7amd** [comment]
+1

**Comment by thomasegriffith** [comment]
Hi @AndreaGriffiths11 is there any update on this?  Thanks in advance!

**Comment by mariusghitoiu** [comment]
+1

**Comment by flaw** [comment]
This is a _huge_ pain in the butt for projects using conventional commits where `fix:` is the most common prefix and does not imply at all that it would resolve any and all issues the PR is related to. If it was limited to having `fix:` in the PR title or body would be one thing, but just having a single commit containing `fix:` in a larger chain of commits triggering this auto-closing is ridiculous.

**Comment by BrutishGuy** [comment]
I also have use-case requiring PRs which mention issues, but do not guarantee the issue is closed/resolved from the PR necessarily, requiring further review/QA. Looking at the many other upvotes and comments, it's a highly requested feature. Adding my voice here to the demand, this would be a widely useful feature.

**Comment by ADTC** [comment]
Need it, seriously. I want to see the issue / PR in the sidebar but sometimes don't want the issue closed if PR is merged.

**Comment by dsilvam** [comment]
+1

**Comment by nollymar** [comment]
+1

**Comment by fmontes** [comment]
+1

**Comment by vorg-code** [comment]
+1

**Comment by yassinotec** [comment]
+1

**Comment by ArturFortunato** [comment]
+1

**Comment by terenty-rezman** [comment]
+1

**Comment by m3nax** [comment]
+1

**Comment by Zadjil-AR** [comment]
There is a cross-over request which is looking for the  change in the opposite direction https://github.com/orgs/community/discussions/13240
This feels like they are 2 sides of the same coin. 
The users of GitHub are looking for ways to control the automatic actions that GitHub do, which we are very thankful they do when appropriate.

**Comment by OnurKlc** [comment]
+1

**Comment by martin-mael** [comment]
+1

**Comment by bitdivine** [comment]
Is there an option to disable all such built-in actions?  We have github actions that we can use to implement desired behaviour, such as closing issues, _if_ it is desired.  There is no need for github to do this by default.

**Comment by Tolu-Mals** [comment]
I need this 🫠

**Comment by mykhaliuk** [comment]
Guys, make this happen please 🙏🏻

**Comment by krickl** [comment]
I also need this! Issues are nearly never done just because a PR is merged. There maybe more PRs, manual testing, documentation, reporting, backports, ... If you are regularly done with an issue by merging a single PR you probably create too many issues.

**Reply by kingthorin** [reply]
Just reference them without using a closing keyword?

https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue

**Reply by krickl** [reply]
I want the PRs listed in the "Development" panel of the issue, for a compact overview at a predictable place on the page. When you only mention the issue number without a closing keyword, the PR will not show up on this panel.

**Reply by kingthorin** [reply]
Oh okay, that makes sense.

**Comment by marcelo-michels** [comment]
I also really need this!

**Comment by lidiaparrilla** [comment]
I doubt this is going anywhere. It is a discussion marked as Answered by @AndreaGriffiths11. I believe we should find the related “feature request”, that she was mentioning, and follow-up there.

**Comment by marcelo-michels** [comment]
@AndreaGriffiths11 any update?

**Comment by jonas-gomes-vs** [comment]
@AndreaGriffiths11  I would really appreciate if this feature would be released.

**Comment by jamesicon** [comment]
We would really like this feature as we use github boards as our kanban system and it would be helpful to show the linked PR without it closing the issue when merged as most issues can take several PR's to fully fix and we don't consider something done until its in production which merging only goes to a staging env

**Comment by giswqs** [comment]
It has been more that three years since this feature request submission. Disappointing to see that to this date this feature still has not been implemented.

**Comment by sax555** [comment]
we are also encountering the same issue.... is there any workaround available?

**Reply by kingthorin** [reply]
Reference them without one of the closing key words.

**Reply by bitdivine** [reply]
That is a workaround but no solution.  This needs to be fixed.

**Comment by Balros** [comment]
+1

**Comment by k-taiga** [comment]
I really need this feature. Please make this happen :pray:

**Comment by lidiaparrilla** [comment]
Guys, I don't think they will be reading a closed discussion. I just opened a new one in hope for attention from the GitHub team: https://github.com/orgs/community/discussions/66741. Go give it some support!

**Comment by KateRezentes** [comment]
We really need this at my work too.

**Comment by yash10019coder** [comment]
We really need this this is crucial.

**Comment by ErichRonis** [comment]
file:///I:/setup.exe

**Comment by intuji-vijayrana** [comment]
Any update on this feature request?

**Reply by lidiaparrilla** [reply]
This feature request is marked as solved by a moderator. I suggest you support this new open one to gather more attention from the dev team: [https://github.com/orgs/community/discussions/66741#discussioncomment-8601663](https://github.com/orgs/community/discussions/66741#discussioncomment-8601663)

**Comment by ryu1** [comment]
+1

**Comment by dvega-flexion** [comment]
Not having this option may have discouraged my team from creating tasks altogether, seems like this could cause a proliferation of untracked branches.

**Comment by mario-set** [comment]
+1

We have created some workflow where we have some "collection" type issue tickets, that we need several steps to complete, each with its own PR.
its really  bad that merging a PR immediatly closes the issue and even moves to `Done` column in any associated project, its a mess. so it would be nice to implement this feature, please.

We have a workflow with "collection"-type issue tickets that require several steps to complete, each with its own PR.
Is a constant problem that merging a PR immediately closes the issue and even moves it to the Done column in any associated project.

It would be nice if this feature gets implemented to address this problem.

**Reply by AlanGreene** [reply]
https://github.blog/changelog/2025-04-23-users-can-now-choose-whether-merging-linked-pull-requests-automatically-closes-the-issue/

**Comment by marc-hb** [comment]
https://github.blog/changelog/2025-04-23-users-can-now-choose-whether-merging-linked-pull-requests-automatically-closes-the-issue/
